src/util/include/util/Color.h

```diff
@@ -11,149 +11,160 @@
+    return is >> rhs.red >> rhs.green >> rhs.blue >> rhs.alpha;
+}
+
 namespace Util {
 
 constexpr auto rgb_to_GdkRGBA(Color color) -> GdkRGBA;
 constexpr auto argb_to_GdkRGBA(Color color) -> GdkRGBA;
 constexpr auto argb_to_GdkRGBA(Color color, double alpha) -> GdkRGBA;
 constexpr auto GdkRGBA_to_argb(const GdkRGBA& color) -> Color;
 constexpr auto GdkRGBA_to_rgb(const GdkRGBA& color) -> Color;
-constexpr auto colorU16_to_argb(const ColorU16& color) -> Color;
 
+constexpr auto ColorU16_to_argb(const ColorU16& color) -> Color;
+constexpr auto argb_to_ColorU16(const Color& color) -> ColorU16;
 constexpr auto GdkRGBA_to_ColorU16(const GdkRGBA& color) -> ColorU16;
 
 
 void cairo_set_source_rgbi(cairo_t* cr, Color color);
 void cairo_set_source_rgbi(cairo_t* cr, Color color, double alpha);
 
-constexpr auto floatToUIntColor(double color) -> uint32_t;
+constexpr auto floatToUIntColor(double color) -> uint8_t;
 
 }  // namespace Util
 
-constexpr auto Util::rgb_to_GdkRGBA(const Color color) -> GdkRGBA {  //
-    return Util::argb_to_GdkRGBA(Color{0xFF000000U | color});
+constexpr auto Util::rgb_to_GdkRGBA(Color color) -> GdkRGBA {  //
+    color.a = 0xFF;
+    return Util::argb_to_GdkRGBA(color);
 }
 
 constexpr auto Util::argb_to_GdkRGBA(const Color color) -> GdkRGBA {
-    return {((color >> 16U) & 0xFFU) / 255.0,  //
-            ((color >> 8U) & 0xFFU) / 255.0,   //
-            ((color >> 0U) & 0xFFU) / 255.0,   //
-            ((color >> 24U) & 0xFFU) / 255.0};
+    return {color.r / 255.0,  //
+            color.g / 255.0,  //
+            color.b / 255.0,  //
+            color.a / 255.0};
 }
 
 constexpr auto Util::argb_to_GdkRGBA(Color color, double alpha) -> GdkRGBA {
-    return {((color >> 16U) & 0xFFU) / 255.0,  //
-            ((color >> 8U) & 0xFFU) / 255.0,   //
-            ((color >> 0U) & 0xFFU) / 255.0,   //
+    return {color.r / 255.0,  //
+            color.g / 255.0,  //
+            color.b / 255.0,  //
             alpha};
 }
 
 constexpr auto Util::GdkRGBA_to_argb(const GdkRGBA& color) -> Color {
-    return floatToUIntColor(color.alpha) << 24U |  //
-           GdkRGBA_to_rgb(color);                  //
+    auto ret = GdkRGBA_to_rgb(color);
+    ret.a = floatToUIntColor(color.alpha);
+    return ret;
 }
 
 constexpr auto Util::GdkRGBA_to_rgb(const GdkRGBA& color) -> Color {
-    return floatToUIntColor(color.red) << 16U |   //
-           floatToUIntColor(color.green) << 8U |  //
-           floatToUIntColor(color.blue);
+    return Color{floatToUIntColor(color.red),    //
+                 floatToUIntColor(color.green),  //
+                 floatToUIntColor(color.blue)};
+}
+
+constexpr auto Util::argb_to_ColorU16(const Color& color) -> ColorU16 {
+    return {static_cast<uint16_t>((color.r << 8U) + color.r), static_cast<uint16_t>((color.g << 8U) + color.g),
+            static_cast<uint16_t>((color.b << 8U) + color.b), static_cast<uint16_t>((color.a << 8U) + color.a)};
```

Comment:
> This way we could keep an invariance for the transformation argb_to_ColorU16(ColorU16_to_argb(c) == c

This is also true for my version: 0x00 => 0x00.00 => 0x00, 0xab => 0xab.ab => 0xab, 0xff => 0xff.ff => 0xff.
In my version, all the color ranges equally map to the other range: 0xff (true black u8) => 0xffff (true black u16)

With your version: 0xff => 0xff00, which is nearly equal to 0xfeff => 0xfe.

Reply:
You're right the invariance is actually not the solid argument.

My confusion came from the fact that I though of converting `8U` to `16U` as of simple multiplication/rescaling by factor `0xff` but it is actually by `0x101`.
I had to do the math in decimal to understand the rational. If you have the numbers `[0-9]` and you want to spread them out over the range `[0-99]` you cannot simply multiply by `10` as you would give a way the whole region of `[91-99]` then (basically loosing contrast).
:exploding_head: That's definitely my TIL. Thanks :pray: 

Might be worth a comment over the function though. Otherwise people as knowledgeable as me in bit operations might think it's unnecessary :sweat_smile:

Reply:
Done